File: cmd/storage.ts
import { error, exec } from "./utils";

// function calcSize(data: number) {
//     const capacity = ["bytes", "KB", "MB", "GB", "TB"];
//     while (data > 1024) {
//         data /= 1024;
//         capacity.shift();
//     }
//     return data.toFixed(2) + capacity[0];
// }

/**
 * Get information concerning the user's storage drive
 */
export default function storage(id: string) {
    const data = exec(`adb -s ${id} shell df /data -h`);    
    if (data.error) throw new Error(data.stderr);
    try {
        const [size, used, available, storage] = data
            .stdout
            .match(/\d+(G|M|K)?%?\s/g)!
        return {
            size: size.trim() + "B",
            used: used.trim() + "B",
            available: available.trim() + "B",
            usage: storage.trim()
        }

    } catch (err) {
        // @ts-expect-error ...
        error(err.message)
        throw new Error("Unable to get storage data")
    }
}

File: go/cli/storage.go
package cli

import (
	"fmt"
	"regexp"
	"strings"
	"roidy/utils"
)

type StorageData struct {
	Size  string // total size in bytes
	Used  string // used space in bytes
	Avail string // available space in bytes
	Usage string // percentage of space used
}

// Get information concerning the user's storage device
func Storage(id string) (StorageData, error) {
	data, err := utils.Execs(fmt.Sprintf("adb -s %s shell df /data -h", id))
	if err != nil {
		return StorageData{}, err
	}
	parts := regexp.MustCompile(`d+(G|M|K)?%?\s`).FindAllString(data, -1)
	return StorageData{
		Size:  strings.TrimSpace(parts[0]) + "B",
		Used:  strings.TrimSpace(parts[1]) + "B",
		Avail: strings.TrimSpace(parts[2]) + "B",
		Usage: strings.TrimSpace(parts[3]),
	}, nil
}
